app/helpers/viewers_helper.rb

```diff
@@ -1,13 +1,54 @@
 # helper for _viewers_indicators partial
 module ViewersHelper
+  private def get_viewer_list(data, link)
+    if data.blank?
+      link ? t('shared.viewers.you_link') : t('shared.viewers.you')
+    else
+      names = data.to_a.map { |id| User.find(id).name }.to_sentence
+      link ? t('shared.viewers.many', viewers: names) : names
+    end
+  end
+
+  private def get_viewers(data, data_type, obj)
+    objs = obj.where(userid: data.userid).all.order('created_at DESC')
+    objs.each do |ob|
+      item = ob.send(data_type)
+      return ob.viewers if item.include?(data.id)
+    end
+  end
+
   def number_of_viewers(current_user_id, strategy_owner_id, viewers)
     if strategy_owner_id == current_user_id
-      viewer_names = viewers.map { |id| User.where(id: id).first.name }
+      names = viewers.map { |id| User.where(id: id).first.name }
       t('shared.viewers_indicator.viewers_html',
         count: viewers.count,
-        names: viewer_names.to_sentence)
+        names: names.to_sentence)
     else
       t('shared.viewers_indicator.you_are_viewer_html', count: viewers.length)
     end
   end
+
+  def viewers_hover(data, link)
+    result = { class: 'yes_title',
+               title: get_viewer_list(data, link).html_safe }
+    if link
+      content_tag(:span, result) do
+        link_to link.name, (link.class.link + link.id.to_s)
+      end
+    else
+      result[:class] += ' small_margin_right'
```

Comment:
this refactor is a really great improvement! I think it would be even better if we could stop generating HTML in the helpers and just have it return data values (e.g. what goes in the `title` attribute) but we can do that later.